src/libcore/lib.rs

```diff
@@ -208,3 +208,84 @@ pub use coresimd::simd;
 #[unstable(feature = "stdsimd", issue = "48556")]
 #[cfg(not(stage0))]
 pub use coresimd::arch;
+
+// FIXME: move to appropriate location
+#[doc(hidden)]
+#[allow(missing_docs)]
+#[unstable(feature = "generic_assert_internals", issue = "44838")]
+pub mod assert_helper {
+    use fmt::{self, Debug, Formatter};
+
+    #[unstable(feature = "generic_assert_internals", issue = "44838")]
+    #[allow(missing_debug_implementations)]
+    pub enum Captured<T> {
+        Value(T),
+        NotCopy,
+        Unevaluated,
+    }
+
+    #[unstable(feature = "generic_assert_internals", issue = "44838")]
+    pub struct DebugFallback<T> {
+        value: Captured<T>,
+        alt: &'static str,
+    }
+
+    impl<T> DebugFallback<T> {
+        #[inline]
+        #[unstable(feature = "generic_assert_internals", issue = "44838")]
+        pub fn new(value: Captured<T>, alt: &'static str) -> Self {
+            DebugFallback { value, alt }
+        }
+    }
+
+    #[unstable(feature = "generic_assert_internals", issue = "44838")]
+    impl<T> Debug for DebugFallback<T> {
+        default fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+            f.write_str(self.alt)
```

Comment:
Is it possible for self.value to be Unevaluated here? If so, I would expect this to print `"(unevaluated)"` even for non-Debug types.

Reply:
Right, this needs to be

```rust
match self.value {
    Value | NotCopy => alt,
    Unevaluated => "(unevaluated)",
}
```